presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java

```diff
@@ -693,4 +697,38 @@ public static long widthBucket(@SqlType(StandardTypes.DOUBLE) double operand, @S
 
         return lower;
     }
+
+    @Description("Cosine similarity between two Map<Varchar, Double>")
+    @ScalarFunction
+    @SqlType(StandardTypes.DOUBLE)
+    public static double cosineSimilarity(@SqlType(MAP_STRING_DOUBLE) Block leftMap, @SqlType(MAP_STRING_DOUBLE) Block rightMap)
+    {
+        double dotProduct = dotProduct(leftMap, rightMap);
+        double normLeftMap = Math.sqrt(dotProduct(leftMap, leftMap));
```

Comment:
Constructing `TypedSet` objects is quite expensive, so you should implement a `magnitude` helper that avoids that to optimize this